layersvt/device_simulation.cpp

```diff
@@ -476,7 +684,32 @@ class JsonLoader {
             GetValue(value, i, &queue_family_properties);
             dest->push_back(queue_family_properties);
         }
-        return dest->size();
+        return static_cast<int>(dest->size());
+    }
+
+    int GetArray(const Json::Value &parent, const char *name, ArrayOfVkFormatProperties *dest) {
+        DebugPrintf("\t\tJsonLoader::GetArray(ArrayOfVkFormatProperties)\n");
+        const Json::Value value = parent[name];
+        if (value.type() != Json::arrayValue) {
+            return -1;
+        }
+        dest->clear();
+        const int count = static_cast<int>(value.size());
+        for (int i = 0; i < count; ++i) {
+            // Get a format structure from JSON.
+            DevsimFormatProperties devsim_format_properties = {};
+            GetValue(value, i, &devsim_format_properties);
+            // Split the JSON-acquired data into VkFormat and VkFormatProperties.
+            const VkFormat format = devsim_format_properties.formatID;
+            VkFormatProperties vk_format_properties = {};
+            vk_format_properties.linearTilingFeatures = devsim_format_properties.linearTilingFeatures;
+            vk_format_properties.optimalTilingFeatures = devsim_format_properties.optimalTilingFeatures;
+            vk_format_properties.bufferFeatures = devsim_format_properties.bufferFeatures;
+            if (IsFormatSupported(vk_format_properties)) {
```

Comment:
Does this check make it impossible to spoof an unsupported format?

Reply:
Devsim tries not to make anything impossible.  You can certainly enable a format that is not consistent with nor supported by the actual implementation.

IsFormatSupported() just checks if the format's 3 featureflags fields are all zero.  [Per the spec that format is unsupported](https://github.com/LunarG/VulkanTools/blob/mikew_devsim120/layersvt/device_simulation.cpp#L442) and Devsim doesn't bother adding to its internal container.

Reply:
But if you don't add it to the internal container, won't devsim use the underlying device's format properties? What if I want to restrict the formats available to an app by specifying an unsupported format in the json?